File: java/rottingOranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        int nRows = grid.length;
        int nCols = grid[0].length;

        Queue<int[]> qu = new LinkedList<>();
        int nFresh = 0;
        int[] delimit = new int[]{-1, -1};
        for (int r = 0; r < nRows; r++)
        {
            for (int c = 0; c < nCols; c++)
            {
                if (grid[r][c] == 1) nFresh++; 
                else if (grid[r][c] == 2)
                {
                    int[] pos = new int[]{r,c};
                    qu.add(pos);
                };
            }
        }
        qu.add(delimit);
        
        int rv = 0;
        int[][] directions = {{-1, 0},{1, 0},{0, -1},{0, 1}};
        while (!qu.isEmpty())
        {
            int[] pos = qu.poll();
            if (pos[0] == -1)
            {
                rv++;
                if (!qu.isEmpty()) qu.add(pos);
            }
            else
            {
                for (int[] d: directions)
                {
                    int rNew = pos[0] + d[0], cNew = pos[1] + d[1];
                    
                    if (!((0 <= rNew && rNew < nRows) &&  
                        (0 <= cNew && cNew < nCols))) continue;
                    else 
                    {
                        if (grid[rNew][cNew] == 1)
                        {
                            grid[rNew][cNew] = 2;
                            nFresh--;
                            int[] newPos = {rNew, cNew};
                            qu.add(newPos);
                        }
                    }
                }
            } 
        }
        if (nFresh == 0) return rv - 1;
        return -1; 
    }
}

File: python/graphs/rottingOranges.py
from typing import List
from collections import deque
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        q = deque()

        n_rows, n_cols = len(grid), len(grid[0])
        fresh_oranges = 0
        for r in range(n_rows):
            for c in range(n_cols):
                if grid[r][c] == 2:
                    q.append((r,c))
                if grid[r][c] == 1:
                    fresh_oranges += 1
        
        q.append((-1, -1))
        rv = 0
        print(q)
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        while q:
            orange = q.popleft()
            r = orange[0]
            c = orange[1]
            # if we find a divider between search layers, append to rv
            if orange == (-1, -1):
                    rv += 1
                    # add another "divider" into the queue if not empty
                    # avoids infinite loop
                    if q:
                        q.append((-1,-1))
            else:
                for d_r, d_c in directions:
                    # if out of bounds, do nothing
                    if (0 <= r + d_r < n_rows) and (0 <= c + d_c < n_cols):
                        # if fresh, change to rotting, and add to queue
                        if grid[r + d_r][c + d_c] == 1:
                            grid[r + d_r][c + d_c] = 2
                            fresh_oranges -= 1
                            q.append((r + d_r, c + d_c))
                            
        return rv
                    
grid = [[2,1,1],[1,1,0],[0,1,1]]
print(Solution().orangesRotting(grid))

